control-plane/connect-inject/endpoints_controller.go

```diff
@@ -609,36 +604,61 @@ func (r *EndpointsController) createGatewayRegistrations(pod corev1.Pod, service
 		meta[MetaKeyConsulWANFederation] = "1"
 	}
 
-	port, err := strconv.Atoi(pod.Annotations[annotationMeshGatewayContainerPort])
-	if err != nil {
-		return nil, err
-	}
-
 	service := &api.AgentService{
-		Kind:    api.ServiceKindMeshGateway,
 		ID:      pod.Name,
-		Service: "mesh-gateway",
-		Port:    port,
 		Address: pod.Status.PodIP,
 		Meta:    meta,
-		TaggedAddresses: map[string]api.ServiceAddress{
-			"lan": api.ServiceAddress{
+	}
+
+	// Set the service values and Consul consulNS based on the kind of gateway.
+	gatewayKind := pod.Annotations[annotationGatewayKind]
+	var consulNS string
+	switch gatewayKind {
+	case "mesh":
+		service.Kind = api.ServiceKindMeshGateway
+		service.Service = "mesh-gateway"
+
+		port, err := strconv.Atoi(pod.Annotations[annotationMeshGatewayContainerPort])
+		if err != nil {
+			return nil, err
+		}
+		service.Port = port
+
+		wanAddr, wanPort, err := r.getWanData(pod, serviceEndpoints)
+		if err != nil {
+			return nil, err
+		}
+		service.TaggedAddresses = map[string]api.ServiceAddress{
+			"lan": {
 				Address: pod.Status.PodIP,
 				Port:    port,
 			},
-			"wan": api.ServiceAddress{
+			"wan": {
 				Address: wanAddr,
 				Port:    wanPort,
 			},
-		},
+		}
+
+		consulNS = r.consulNamespace(pod.Namespace)
+	case "terminating":
+		service.Kind = api.ServiceKindTerminatingGateway
+		service.Service = "terminating-gateway"
+		service.Port = 8443
+
+		if ns, ok := pod.Annotations[annotationTerminatingGatewayNamespace]; ok {
+			consulNS = ns
+		}
 	}
+	service.Namespace = consulNS
```

Comment:
nit: I think this would be a bit simpler if you set that in the terminating gw case because then you don't need the extra variable.